rancher/rancher
# [v2.7] Better handling of system-default-registry flag for RKE2/K3s

## Issue: https://github.com/rancher/rancher/issues/38890
 
## Problem
Handling of the `system-default-registry` flag resulted in cluster-level private registries not being honored. Prior PR somewhat addressed this, however edge cases were found. This PR addresses the failing test case wherein an unauthenticated global level registry is configured as well as an unauthenticated cluster level registry.
 
## Solution
While working on #38897 I found a more appropriate solution where we can always pass the flag, ensuring all configurations of cluster level private registries and global level private registries work as expected.
 
## Testing
Test case 1) 

1. Create 1 unauthenticated and 1 authenticated registry containing all system images for a released version of Rancher (Images for registries are generated with reach release and RC, and can be located in the `rancher-images.txt` file.) 
  1a. The unauthenticated registry will act as your global level registry 
  1b. The authenticated registry will act as your cluster level registry 
2. Change the global `system-default-registry` in the Rancher global settings to point to your unauthenticated registry 
3. Begin to create an RKE2/K3s cluster, ensuring you chose the correct version of k8's for the images loaded into your registries
4. In the `Cluster Configuration` section, select `Registries`
5. The `Pull images for Rancher from a private registry` option should already be selected, and the value should default to your global level registry. Change the value to point to your cluster level registry.
6. For `Authentication` select `Create a basic HTTP Auth Secret` and pass in the username and password of your authenticated registry 
7. Create the cluster and wait for it to become available 
8. Navigate to the `Cluster Explorer`
9. In the top right corner, select all namespaces 
10. Navigate to the `pods` section of the explorer so that all pods are visible
11. Check that all images pulled by default (system-default-images) are pulled from the cluster level registry and not the global level registry. 
12. It is expected that the `fleet-agent` image is not pulled from the cluster level registry, this is a known issue that was intentionally not resolved in this PR 

Test Case 2) 

1. Create 2 unauthenticated registries containing all system images for a released version of Rancher (Images for registries are generated with reach release and RC, and can be located in the `rancher-images.txt` file.) 
  1a. One unauthenticated registry will act as your global level registry 
  1b. The other unauthenticated registry will act as your cluster level registry 
2. Change the global `system-default-registry` in the Rancher global settings to point to your unauthenticated registry 
3. Begin to create an RKE2/K3s cluster, ensuring you chose the correct version of k8's for the images loaded into your registries
4. In the `Cluster Configuration` section, select `Registries`
5. The `Pull images for Rancher from a private registry` option should already be selected, and the value should default to your global level registry. Change the value to point to your cluster level registry. Do not provide any authentication. 
6. Create the cluster and wait for it to become available 
7. Navigate to the `Cluster Explorer`
8. In the top right corner, select all namespaces 
9. Navigate to the `pods` section of the explorer so that all pods are visible
10. Check that all images pulled by default (system-default-images) are pulled from the cluster level registry and not the global level registry. 
11. It is expected that the `fleet-agent` image is not pulled from the cluster level registry, this is a known issue that was intentionally not resolved in this PR 


## Engineering Testing
### Manual Testing
I've done both of the above test scenarios with this change. Test case 2, which was previously failing, is now working and most system images are being pulled from the unauthenticated cluster level private registry as expected. I only have a single unauthenticated registry to test with, however setting the global system-default-registry to `docker.io` allowed me to test the failing scenario. 

### Automated Testing
No automated testing has been added

## QA Testing Considerations
The test plan previously created for this ticket can be reused.
 
### Regressions Considerations
All combinations of global and cluster level registries should be tested.
